File: Example/PrebidDemo/PrebidDemoObjectiveC/Examples/GAM/OriginalAPI/GAMOriginalAPIDisplayBannerViewController.m
/*   Copyright 2019-2022 Prebid.org, Inc.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#import "GAMOriginalAPIDisplayBannerViewController.h"
#import "PrebidDemoMacros.h"

@import PrebidMobile;
@import GoogleMobileAds;

NSString * const storedImpDisplayBanner = @"prebid-demo-banner-320-50";
NSString * const gamAdUnitDisplayBannerOriginal = @"/21808260008/prebid_demo_app_original_api_banner";

@interface GAMOriginalAPIDisplayBannerViewController ()

// Prebid
@property (nonatomic) BannerAdUnit *adUnit;

// GAM
@property (nonatomic) GAMBannerView *gamBanner;

@end

@implementation GAMOriginalAPIDisplayBannerViewController

- (void)loadView {
    [super loadView];
    
    [self createAd];
}

- (void)dealloc {
    // Clean up global ORTB config
    [Targeting.shared setGlobalORTBConfig:nil];
}

- (void)createAd {
    // 1. Set global ORTB
    NSString *globalORTB = [NSString stringWithFormat:
        @"{"
            "\"ext\": {"
                "\"myext\": {"
                    "\"test\": 1"
                "}"
            "},"
            "\"displaymanager\": \"Google\","
            "\"displaymanagerver\": \"%@\""
        "}",
        GADGetStringFromVersionNumber([GADMobileAds sharedInstance].versionNumber)
    ];
    
    [Targeting.shared setGlobalORTBConfig:globalORTB];
    
    // 2. Create a BannerAdUnit
    self.adUnit = [[BannerAdUnit alloc] initWithConfigId:storedImpDisplayBanner size:self.adSize];
    
    // 3. Configure banner parameters
    BannerParameters * parameters = [[BannerParameters alloc] init];
    parameters.api = @[PBApi.MRAID_2];
    self.adUnit.bannerParameters = parameters;
    
    // 4. Set impression-level ORTB
    [self.adUnit setImpORTBConfig:@"{\"bidfloor\":0.01,\"banner\":{\"battr\":[1,2,3,4]}}"];
    
    // 5. Create a GAMBannerView
    GAMRequest * gamRequest = [GAMRequest new];
    self.gamBanner = [[GAMBannerView alloc] initWithAdSize:GADAdSizeFromCGSize(self.adSize)];
    self.gamBanner.adUnitID = gamAdUnitDisplayBannerOriginal;
    self.gamBanner.rootViewController = self;
    self.gamBanner.delegate = self;
    
    // Add GMA SDK banner view to the app UI
    [self.bannerView addSubview:self.gamBanner];
    
    // 6. Make a bid request to Prebid Server
    @weakify(self);
    [self.adUnit fetchDemandWithAdObject:gamRequest completion:^(enum ResultCode resultCode) {
        @strongify(self);
        if (!self) { return; }
        
        // 7. Load GAM Ad
        [self.gamBanner loadRequest:gamRequest];
    }];
}

// MARK: - GADBannerViewDelegate

- (void)bannerViewDidReceiveAd:(GADBannerView *)bannerView {
    [AdViewUtils findPrebidCreativeSize:bannerView success:^(CGSize size) {
        [self.gamBanner resize:GADAdSizeFromCGSize(size)];
    } failure:^(NSError * _Nonnull error) {
        PBMLogError(@"%@", error.localizedDescription)
    }];
}

- (void)bannerView:(GADBannerView *)bannerView didFailToReceiveAdWithError:(NSError *)error {
    PBMLogError(@"%@", error.localizedDescription)
}

@end

File: Example/PrebidDemo/PrebidDemoSwift/Examples/GAM/OriginalAPI/GAMOriginalAPIDisplayBannerViewController.swift
/*   Copyright 2019-2022 Prebid.org, Inc.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

import UIKit
import PrebidMobile
import GoogleMobileAds

fileprivate let storedImpDisplayBanner = "prebid-demo-banner-320-50"
fileprivate let gamAdUnitDisplayBannerOriginal = "/21808260008/prebid_demo_app_original_api_banner"

class GAMOriginalAPIDisplayBannerViewController:
    BannerBaseViewController,
    GoogleMobileAds.BannerViewDelegate {
    
    // Prebid
    private var adUnit: BannerAdUnit!
    
    // GAM
    private var gamBanner: AdManagerBannerView!
    
    override func loadView() {
        super.loadView()
        
        createAd()
    }
    
    deinit {
        // Clean up global ORTB config
        Targeting.shared.setGlobalORTBConfig(nil)
    }
    
    func createAd() {
        // 1. Set global ORTB
        let globalORTB = """
        {
          "ext": {
            "myext": {
              "test": 1
            }
          },
          "displaymanager": "Google",
          "displaymanagerver": "\(string(for: MobileAds.shared.versionNumber))"
        }
        """
        Targeting.shared.setGlobalORTBConfig(globalORTB)
        
        // 2. Create a BannerAdUnit
        adUnit = BannerAdUnit(configId: storedImpDisplayBanner, size: adSize)
        adUnit.setAutoRefreshMillis(time: 30000)
        
        // 3. Configure banner parameters
        let parameters = BannerParameters()
        parameters.api = [Signals.Api.MRAID_2]
        adUnit.bannerParameters = parameters
        
        // 4. Set impression-level ORTB
        adUnit.setImpORTBConfig("{\"bidfloor\":0.01,\"banner\":{\"battr\":[1,2,3,4]}}")
        
        // 5. Create a GAMBannerView
        gamBanner = AdManagerBannerView(adSize: adSizeFor(cgSize: adSize))
        gamBanner.adUnitID = gamAdUnitDisplayBannerOriginal
        gamBanner.rootViewController = self
        gamBanner.delegate = self
        
        // Add GMA SDK banner view to the app UI
        bannerView?.addSubview(gamBanner)
        
        // 6. Make a bid request to Prebid Server
        let gamRequest = AdManagerRequest()
        adUnit.fetchDemand(adObject: gamRequest) { [weak self] resultCode in
            PrebidDemoLogger.shared.info("Prebid demand fetch for GAM \(resultCode.name())")
            
            // 7. Load GAM Ad
            self?.gamBanner.load(gamRequest)
        }
    }
    
    // MARK: - GADBannerViewDelegate
    
    func bannerViewDidReceiveAd(_ bannerView: GoogleMobileAds.BannerView) {
        AdViewUtils.findPrebidCreativeSize(bannerView, success: { size in
            guard let bannerView = bannerView as? AdManagerBannerView else { return }
            bannerView.resize(adSizeFor(cgSize: size))
        }, failure: { (error) in
            PrebidDemoLogger.shared.error("Error occuring during searching for Prebid creative size: \(error)")
        })
    }
    
    func bannerView(
        _ bannerView: GoogleMobileAds.BannerView,
        didFailToReceiveAdWithError error: Error
    ) {
        PrebidDemoLogger.shared.error("GAM did fail to receive ad with error: \(error)")
    }
}
